packages/java-parser/src/productions/expressions.js

```diff
@@ -604,21 +614,19 @@ function defineRules($, t) {
     return true;
   });
 
-  let firstForUnaryExpressionNotPlusMinus = undefined;
   $.RULE("isReferenceTypeCastExpression", () => {
-    if (firstForUnaryExpressionNotPlusMinus === undefined) {
-      const firstUnaryExpressionNotPlusMinus = this.computeContentAssist(
-        "unaryExpressionNotPlusMinus",
-        []
-      );
-      const nextTokTypes = firstUnaryExpressionNotPlusMinus.map(
-        x => x.nextTokenType
-      );
-      // uniq
-      firstForUnaryExpressionNotPlusMinus = nextTokTypes.filter(
-        (v, i, a) => a.indexOf(v) === i
-      );
-    }
+    const firstUnaryExpressionNotPlusMinus = this.computeContentAssist(
```

Comment:
we dont want to compute this data every time! it is a slow operation

Reply:
concerning this one, I tried to keep it outside of the isReferenceTypeCastExpression rule, but it was not giving me the correct infos... So I'm not sure what to do here

Reply:
For instance, when trying to parse
 ```java
PlayerDetails other = (PlayerDetails) obj;
```
with `localVariableDeclarationStatement` as entryPoint in `single-sample-runner.js`. I got this:

```
/home/cdessoude/Documents/workplace/prettier-java/packages/java-parser/src/index.js:44
    throw Error(
    ^

Error: Sad sad panda, parsing errors detected in line: 2, column: 39!
Expecting --> ';' <-- but found --> 'obj' <--!
	->localVariableDeclarationStatement
    at Object.parse (/home/cdessoude/Documents/workplace/prettier-java/packages/java-parser/src/index.js:44:11)
    at Object.<anonymous> (/home/cdessoude/Documents/workplace/prettier-java/packages/java-parser/scripts/single-sample-runner.js:12:16)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:831:12)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:622:3)
```

When I print the `firstForUnaryExpressionNotPlusMinus` before consuming the `t.LBrace`, I only have a `UnaryPrefixOperatorNotPlusMinus` and a `UnarySuffixOperator` when I had much more tokens with my change

Reply:
I'm not sure how to proceed on this one: I tried to extract the logic into one method and call it at the end of the parser constructor, but didn't manage to make it work. If I extract the logic, I got this error `Error: Rule ->unaryExpressionNotPlusMinus<- does not exist in this grammar.`, so I guess I am not doing the right thing 😄

Reply:
```javascript
    // ...
    // prevents runtime error during the recording phase
    this.firstForUnaryExpressionNotPlusMinus= [];
    this.performSelfAnalysis();

    const firstUnaryExpressionNotPlusMinus = this.computeContentAssist(
      "unaryExpressionNotPlusMinus",
      []
    );
    const nextTokTypes = firstUnaryExpressionNotPlusMinus.map(
      x => x.nextTokenType
    );
    // uniq
    this.firstForUnaryExpressionNotPlusMinus = nextTokTypes.filter(
      (v, i, a) => a.indexOf(v) === i
    );
```

```javascript
  $.RULE("isReferenceTypeCastExpression", () => {
    $.CONSUME(t.LBrace);
    $.SUBRULE($.referenceType);
    $.MANY(() => {
      $.SUBRULE($.additionalBound);
    });
    $.CONSUME(t.RBrace);
    const firstTokTypeAfterRBrace = this.LA(1).tokenType;

    return (
      this.firstForUnaryExpressionNotPlusMinus.find(tokType =>
        tokenMatcher(firstTokTypeAfterRBrace, tokType)
      ) !== undefined
    );
  });
```